src/hera/workflows/script.py

```diff
@@ -146,44 +151,11 @@ def _constructor_validate(cls, values):
         assert isinstance(constructor, ScriptConstructor)
         return constructor.transform_values(cls, values)
 
-    def _build_inputs(self) -> Optional[ModelInputs]:
-        inputs = super()._build_inputs()
-        func_parameters: List[Parameter] = []
-        func_artifacts: List[Artifact] = []
-        if callable(self.source):
-            if global_config.experimental_features["script_annotations"]:
-                func_parameters, func_artifacts = _get_parameters_and_artifacts_from_callable(self.source)
-            else:
-                func_parameters = _get_parameters_from_callable(self.source)
-
-        if inputs is None and not func_parameters and not func_artifacts:
-            return None
-        elif not func_parameters and not func_artifacts:
-            return inputs
-        elif inputs is None:
-            inputs = ModelInputs(
-                parameters=func_parameters or None, artifacts=[a._build_artifact() for a in func_artifacts] or None
-            )
-
-        already_set_params = {p.name for p in inputs.parameters or []}
-        already_set_artifacts = {p.name for p in inputs.artifacts or []}
-
-        for param in func_parameters:
-            if param.name not in already_set_params and param.name not in already_set_artifacts:
-                if inputs.parameters is None:
-                    inputs.parameters = []
-                inputs.parameters.append(param)
-
-        for artifact in func_artifacts:
-            if artifact.name not in already_set_params and artifact.name not in already_set_artifacts:
-                if inputs.artifacts is None:
-                    inputs.artifacts = []
-                inputs.artifacts.append(artifact._build_artifact())
-
-        return inputs
-
     def _build_template(self) -> _ModelTemplate:
         assert isinstance(self.constructor, ScriptConstructor)
+        hera_outputs_used = self._check_hera_outputs_used()
+        if hera_outputs_used:
+            self._create_new_volume()
```

Comment:
These could be called within `_build_script` as long as we call it first here, i.e.
```suggestion
        script=self._build_script()
```

Then in `_build_script` check/call the `self._create_new_volume` before `env = self._build_env()`